dynamic_programming/tree_height.cpp

```diff
@@ -0,0 +1,59 @@
+// C++ Program to find height of the tree using bottom - up DP.
+
+/*
+ * Given a rooted tree with node 1.
+ * Task is to find the height of the tree.
+ * Example: -
+ * 4
+ * 1 2
+ * 1 3
+ * 2 4
+ * Height of the tree : - 3
+*/
+
+#include<iostream>
+#include<vector>
+
+// global declarations
+// no of nodes max limit.
+const int MAX = 1e5;
+// adjacency list
+std::vector<int> adj[MAX];
+std::vector<bool> visited;
+std::vector<int> dp;
+
+void dp_with_dfs(int u) {
+    visited[u] = true;
+    int child_height = 1;
+    for (int v : adj[u]) {
+        if (!visited[v]) {
+            dp_with_dfs(v);
+
+            // select maximum sub-tree height from all children.
+            child_height = std::max(child_height, dp[v]+1);
+        }
+    }
+    // assigned the max child height to current visited node.
+    dp[u] = child_height;
```

Comment:
What is a `dp` and why do I have to guess?

Reply:
The general way to write a dp equation is termed as dp. Which means "Dynamic Programming".